src/path.cr

```diff
@@ -0,0 +1,847 @@
+          # . element
+          reader.next_char
+        elsif char == '.' && reader.next_char == '.' && (reader.pos + 1 == @name.bytesize || @type.separators.includes?(reader.peek_next_char))
+          # .. element: remove to last /
+          reader.next_char
+          if str.bytesize > dotdot
+            str.back 1
+            while str.bytesize > dotdot && !separator_ord.includes?((str.buffer + str.bytesize).value)
+              str.back 1
+            end
+          elsif !rooted
+            if str.bytesize > 0
+              str << @type.separator
+            end
+            str << ".."
+            dotdot = str.bytesize
+          end
+        else
+          reader.pos = curr_pos # make sure to reset lookahead used in previous condition
+
+          # real path element
+          # add slash if needed
+
+          if rooted && str.bytesize != 1 || !rooted && str.bytesize != 0
+            str << @type.separator
+          end
+
+          loop do
+            str << char
+            char = reader.next_char
+            break if @type.separators.includes?(char) || char == Char::ZERO
+          end
+        end
+      end
+
+      if str.empty?
+        str << '.'
+      end
+
+      last_char = (str.buffer + str.bytesize - 1).value.unsafe_chr
+
+      if add_separator_at_end && !@type.separators.includes?(last_char)
+        str << @type.separator
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Yields each component of this path as a `String`.
+  def each_part
+    @type.each_part(@name) do |part|
+      yield part
+    end
+  end
+
+  # Returns the components of this path as an `Array(String)`.
+  def parts : Array(String)
+    parts = [] of String
+    each_part do |part|
+      parts << part
+    end
+    parts
+  end
+
+  # Converts this path to a Windows path.
+  def to_windows : Path
+    return self if windows?
+
+    new_instance(@name, Type::WINDOWS)
+  end
+
+  # Converts this path to a  POSIX path.
+  def to_posix : Path
+    return self if posix?
+
+    new_instance(@name.gsub(Type::WINDOWS.separator, Type::POSIX.separator), Type::POSIX)
+  end
+
+  # Converts this path to an absolute path. Relative paths are
+  # referenced from the current working directory of the process (`Path.current`)
+  # unless *base* is given, in which case it will be used as the reference path.
+  #
+  # ```
+  # Path["foo"].expand             # => Path["/current/path/foo"]
+  # Path["~/crystal/foo"].expand   # => Path["/home/crystal/foo"]
+  # Path["baz"].expand("/foo/bar") # => Path["/foo/bar/baz"]
+  # ```
+  def expand(base : Path | String = Path.current, home = Path.home) : Path
+    name = @name
+
+    if name.starts_with?('~')
+      home_ends_with_separator = @type.ends_with_separator?(home.@name)
+      home = new_instance(home.@name).normalize.to_s
+      # home = home + @type.separator if home_ends_with_separator && home.size > 1
```

Comment:
Why is this commented? The check above is no longer used either.

Reply:
I can't remember exactly, but I guess it's just left over from before when `normalize` would always chop of the final separator. Needs to be removed.